File: figaro-worker/src/worker/executor.ts
import {
  query,
  type PermissionResult,
  type PermissionMode,
} from "@anthropic-ai/claude-agent-sdk";

import type { NatsClient } from "../nats/client";
import type { TaskPayload } from "../types";
import { serializeMessage } from "../shared/serialize";
import { LoopDetectionSession, buildLoopDetectionHooks, type LoopDetectionConfig } from "../shared/loop-detection";
import { traced } from "../tracing/tracer";
import { HelpRequestHandler } from "./help-request";
import { formatTaskPrompt, WORKER_SYSTEM_PROMPT } from "./prompt-formatter";
import { createDesktopToolsServer } from "./tools";

export class TaskExecutor {
  private client: NatsClient;
  private model: string;
  private claudeCodePath: string | undefined;
  private loopDetection: LoopDetectionConfig | undefined;
  private helpHandler: HelpRequestHandler;
  private currentTaskId: string | null = null;
  private abortController: AbortController | null = null;

  constructor(
    client: NatsClient,
    model: string = "claude-opus-4-6",
    claudeCodePath?: string,
    loopDetection?: LoopDetectionConfig,
  ) {
    this.client = client;
    this.model = model;
    this.claudeCodePath = claudeCodePath;
    this.loopDetection = loopDetection;
    this.helpHandler = new HelpRequestHandler(client);
  }

  async handleTask(payload: TaskPayload): Promise<void> {
    await traced("worker.handle_task", async (span) => {
      const taskId = payload.task_id;
      const prompt = payload.prompt ?? "";
      const optionsDict = (payload.options ?? {}) as Record<string, unknown>;

      if (!taskId) {
        console.error("[executor] Received task without task_id");
        return;
      }

      span.setAttribute("task.id", taskId);
      console.log(`[executor] Executing task ${taskId}: ${prompt.slice(0, 50)}...`);

      try {
        await this.executeTask(taskId, prompt, optionsDict);
      } catch (e) {
        if (this.abortController?.signal.aborted) {
          console.log(`[executor] Task ${taskId} was stopped`);
          return;
        }
        const errMsg = e instanceof Error ? e.message : String(e);
        console.error(`[executor] Task ${taskId} failed: ${errMsg}`);
        await this.client.publishTaskError(taskId, errMsg);
      } finally {
        this.currentTaskId = null;
        this.abortController = null;
      }
    });
  }

  stopTask(taskId: string): boolean {
    if (this.currentTaskId === taskId && this.abortController) {
      console.log(`[executor] Stopping task ${taskId}`);
      this.abortController.abort();
      return true;
    }
    return false;
  }

  private async executeTask(
    taskId: string,
    prompt: string,
    optionsDict: Record<string, unknown>,
  ): Promise<void> {
    await traced("worker.execute_task", async (span) => {
    this.currentTaskId = taskId;
    span.setAttribute("task.id", taskId);

    const startUrl = optionsDict.start_url as string | undefined;
    const formattedPrompt = await traced("worker.format_task_prompt", async () => {
      return formatTaskPrompt(prompt, startUrl);
    });

    const { server: desktopTools, destroySession } = createDesktopToolsServer(this.client);

    const canUseTool = async (
      toolName: string,
      inputData: Record<string, unknown>,
    ): Promise<PermissionResult> => {
      if (toolName !== "AskUserQuestion") {
        return { behavior: "allow", updatedInput: inputData };
      }

      const questions = (inputData.questions ?? []) as Array<Record<string, unknown>>;

      if (!this.currentTaskId) {
        console.warn("[executor] AskUserQuestion: no current task ID");
        return { behavior: "deny", message: "No active task context" };
      }

      console.log(`[executor] AskUserQuestion: ${questions.length} question(s)`);

      const answers = await this.helpHandler.requestHelp(
        this.currentTaskId,
        questions,
        null,
        1800,
      );

      if (answers) {
        console.log("[executor] Received human response for AskUserQuestion");
        return {
          behavior: "allow",
          updatedInput: {
            questions,
            answers,
          },
        };
      } else {
        console.warn("[executor] Timeout or error waiting for human response");
        return {
          behavior: "deny",
          message: "Timeout waiting for human response. Proceed with your best judgment.",
        };
      }
    };

    const permissionMode = (optionsDict.permission_mode as PermissionMode | undefined) ?? "bypassPermissions";

    const hooks = this.loopDetection?.enabled
      ? buildLoopDetectionHooks(
          new LoopDetectionSession(this.loopDetection),
          this.loopDetection,
        )
      : undefined;

    this.abortController = new AbortController();
    const abortController = this.abortController;

    const q = query({
      prompt: formattedPrompt,
      options: {
        systemPrompt: {
          type: "preset" as const,
          preset: "claude_code" as const,
          append: WORKER_SYSTEM_PROMPT,
        },
        permissionMode,
        allowDangerouslySkipPermissions: permissionMode === "bypassPermissions",
        maxTurns: optionsDict.max_turns as number | undefined,
        model: this.model,
        settings: {
          autoMemoryEnabled: false
        },
        settingSources: ["user", "project"],
        canUseTool,
        disallowedTools: ["Memory"],
        mcpServers: { desktop: desktopTools },
        abortController,
        ...(this.claudeCodePath ? { pathToClaudeCodeExecutable: this.claudeCodePath } : {}),
        ...(hooks ? { hooks } : {}),
      },
    });

    let resultMessage: Record<string, unknown> | null = null;

    try {
      for await (const message of q) {
        const serialized = serializeMessage(message);
        serialized.task_id = taskId;

        await this.client.publishTaskMessage(taskId, serialized);

        if ((message as unknown as Record<string, unknown>).type === "result") {
          resultMessage = serialized;
        }
      }
    } finally {
      q.close();
      destroySession();
      if (!abortController.signal.aborted) {
        setTimeout(() => {
          if (!abortController.signal.aborted) {
            console.warn(`[executor] [${taskId.slice(0, 8)}] Force-aborting query subprocess`);
            abortController.abort();
          }
        }, 5_000);
      }
    }

    await this.client.publishTaskComplete(taskId, resultMessage);
    console.log(`[executor] Task ${taskId} completed`);
    });
  }
}

File: legacy/figaro-worker/src/figaro_worker/worker/executor.py
import dataclasses
import logging
from typing import Any

from claude_agent_sdk import ClaudeAgentOptions, ClaudeSDKClient
from claude_agent_sdk._errors import MessageParseError
from claude_agent_sdk.types import (
    PermissionResultAllow,
    PermissionResultDeny,
    ResultMessage,
    ToolPermissionContext,
)

from .client import NatsClient
from .help_request import HelpRequestHandler
from .prompt_formatter import format_task_prompt
from .tools import create_desktop_tools_server

logger = logging.getLogger(__name__)

def serialize_message(msg: Any) -> dict[str, Any]:
    """Serialize a claude-agent-sdk message to a dict."""
    if dataclasses.is_dataclass(msg) and not isinstance(msg, type):
        result = dataclasses.asdict(msg)
        result["__type__"] = type(msg).__name__
        return result
    elif isinstance(msg, dict):
        return msg
    else:
        return {"value": str(msg), "__type__": type(msg).__name__}


class TaskExecutor:
    def __init__(self, client: NatsClient, model: str = "claude-opus-4-6") -> None:
        self.client = client
        self.model = model
        self.help_handler = HelpRequestHandler(client)
        self._current_task_id: str | None = None

    async def handle_task(self, payload: dict[str, Any]) -> None:
        task_id = payload.get("task_id")
        prompt = payload.get("prompt", "")
        options_dict = payload.get("options", {})

        if not task_id:
            logger.error("Received task without task_id")
            return

        logger.info(f"Executing task {task_id}: {prompt[:50]}...")

        # Note: Status is managed by the orchestrator. It sets BUSY when claiming
        # a worker for a task, and sets IDLE when task_complete/error is received.
        # We don't send status messages here to avoid race conditions.

        try:
            await self._execute_task(task_id, prompt, options_dict)
        except Exception as e:
            logger.exception(f"Task {task_id} failed: {e}")
            await self.client.publish_task_error(task_id, str(e))
        finally:
            self._current_task_id = None

    async def _can_use_tool(
        self,
        tool_name: str,
        input_data: dict[str, Any],
        context: ToolPermissionContext,
    ) -> PermissionResultAllow | PermissionResultDeny:
        """Handle tool permission requests.

        Intercepts AskUserQuestion to route questions to humans via NATS
        (UI/Telegram). All other tools are auto-approved.

        See: https://platform.claude.com/docs/en/agent-sdk/user-input
        """
        if tool_name != "AskUserQuestion":
            return PermissionResultAllow(updated_input=input_data)

        questions = input_data.get("questions", [])

        if not self._current_task_id:
            logger.warning("AskUserQuestion: no current task ID")
            return PermissionResultDeny(message="No active task context")

        logger.info(f"AskUserQuestion: {len(questions)} question(s)")

        # Request human help via orchestrator (Telegram/UI)
        answers = await self.help_handler.request_help(
            task_id=self._current_task_id,
            questions=questions,
            timeout_seconds=1800,  # 30 minute timeout
        )

        if answers:
            logger.info("Received human response for AskUserQuestion")
            return PermissionResultAllow(
                updated_input={
                    "questions": questions,
                    "answers": answers,
                }
            )
        else:
            logger.warning("Timeout or error waiting for human response")
            return PermissionResultDeny(
                message="Timeout waiting for human response. Proceed with your best judgment."
            )

    async def _execute_task(
        self,
        task_id: str,
        prompt: str,
        options_dict: dict[str, Any],
    ) -> None:
        # Track current task for AskUserQuestion handling
        self._current_task_id = task_id

        # Format the prompt with XML structure and clarification instructions
        start_url = options_dict.get("start_url")
        formatted_prompt = format_task_prompt(prompt, start_url)

        logger.debug(f"Formatted prompt for task {task_id}")

        desktop_tools = create_desktop_tools_server()

        options = ClaudeAgentOptions(
            permission_mode=options_dict.get("permission_mode", "bypassPermissions"),
            max_turns=options_dict.get("max_turns"),
            model=self.model,
            setting_sources=["user", "project"],
            can_use_tool=self._can_use_tool,
            mcp_servers={"desktop": desktop_tools},
        )

        result_message = None

        try:
            async with ClaudeSDKClient(options) as client:
                await client.query(formatted_prompt)

                async for message in client.receive_response():
                    serialized = serialize_message(message)
                    serialized["task_id"] = task_id

                    await self.client.publish_task_message(task_id, serialized)

                    if isinstance(message, ResultMessage):
                        result_message = serialized
        except MessageParseError as e:
            logger.warning(f"SDK message parse error during task {task_id} (completing with partial results): {e}")

        await self.client.publish_task_complete(task_id, result_message)
        logger.info(f"Task {task_id} completed")
